### PR TITLE
refactor(opsem): encapsulate solver in bv2 ctx

### PR DESCRIPTION
**Pull request by priyasiddharth**
1. keep track of assertions during single vacuity and incremental check
2. add printing of debug location if available

### CODE REVIEW

File: lib/seahorn/BvOpSem2.cc [hunk 1]
```diff
@@ -157,7 +159,6 @@ struct OpSemVisitorBase {
       : m_ctx(ctx), m_efac(m_ctx.getExprFactory()), m_sem(sem) {
 
     trueE = m_ctx.m_trueE;
-    falseE = m_ctx.m_falseE;
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:160]: why?

File: lib/seahorn/BvOpSem2.cc [hunk 2]
```diff
@@ -185,6 +186,7 @@ struct OpSemVisitorBase {
 
     if (const Constant *cv = dyn_cast<Constant>(&v)) {
       return m_ctx.getConstantValue(*cv);
+      falseE = m_ctx.m_falseE;
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:189]: why? won't execute after return

File: lib/seahorn/BvOpSem2.cc [hunk 3]
```diff
@@ -579,59 +581,85 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+                 << "]";
+          } else { INFO << "Assertion passed"; });
     }
   }
+  void visitAssertIf(CallSite CS) {
+    // NOTE: sea.assert.if.not is not supported
+    auto I = CS.getInstruction();
+    Expr ante = lookup(*CS.getArgument(0));
+    Expr conseq = lookup(*CS.getArgument(1));
+    const llvm::DebugLoc &dloc = I->getDebugLoc();
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:645]: what happens if I had no debug location?

File: lib/seahorn/BvOpSem2.cc [hunk 4]
```diff
@@ -579,59 +581,85 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
...
+    Expr e = boolop::lneg(conseq);
     Stats::resume("opsem.incbmc");
     tribool conseqRes = solveWithConstraints(e);
     Stats::stop("opsem.incbmc");
     if (conseqRes) {
-      LOG("opsem", ERR << "Consequent is sat/unknown: " << *I << "\n");
+      LOG(
+          "opsem",
+          if (dloc) {
+            ERR << "Consequent is sat/unknown: "
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:627]: this is only reachable when Sat but prints that might be unknown

File: lib/seahorn/BvOpSem2.cc [hunk 5]
```diff
@@ -2254,6 +2282,22 @@ std::pair<char *, unsigned>
...
   return m_memManager->getGlobalVariableInitValue(gv);
 }
+
+void Bv2OpSemContext::resetSolver() {
+  m_z3_solver->reset();
+  m_addedToSolver.clear();
+}
+
+void Bv2OpSemContext::addToSolver(const Expr e) {
+  if (m_addedToSolver.find(e) == m_addedToSolver.end()) {
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:2292]: use count() instead of find()

File: lib/seahorn/BvOpSem2Context.hh [hunk 6]
```diff
@@ -102,6 +102,7 @@ private:
   std::shared_ptr<ZSolver<EZ3>> m_z3_solver;
 
   bool m_shouldSimplify = false;
+  std::set<Expr> m_addedToSolver;
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2Context.hh:105]: probably unordered set is better here

Comment (agurfinkel) [reply lib/seahorn/BvOpSem2Context.hh:105]: or use ExprSet if don't want to make decision

File: lib/seahorn/BvOpSem2.cc (hunk 3 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:645]: It just prints "Antecedent/Consequest is ..unsat/sat" without debugLoc/instruction info.

File: lib/seahorn/BvOpSem2.cc (hunk 5 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:2292]: done

File: lib/seahorn/BvOpSem2.cc (hunk 3 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2.cc:645]: ok, thought it might crash. looks good now.

File: lib/seahorn/BvOpSem2Context.hh (hunk 6 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2Context.hh:105]: good point

File: lib/seahorn/BvOpSem2.cc (hunk 1 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2.cc:160]: I think this was deleted by mistake, and the place where the line is moved is dead code.

File: lib/seahorn/BvOpSem2.cc (hunk 2 above)
Comment (agurfinkel) [reply lib/seahorn/BvOpSem2.cc:189]: this is dead code. I think it is a typo. See above.

File: lib/seahorn/BvOpSem2.cc (hunk 4 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: good point. fixed.

File: lib/seahorn/BvOpSem2.cc (hunk 1 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: yes.. this is a mix of editor clang-format/misbehaving and my eyeballs not catching it.

File: lib/seahorn/BvOpSem2.cc (hunk 2 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc]: same as above.

**Comment by priyasiddharth** [comment]
Should be ready for another review. thanks

**Comment by priyasiddharth** [comment]
Ok I moved all references to a `solver` from `visitAssert` to `solveWithConstraints`.  This makes `solveWithConstraints` reusable. It can be called in fresh/incremental modes.

**Pull request merged**